Added kickoff exercise and grader

ViliamVadocz opened the pull request:
Includes default kickoffs but it should be easy to customise. Also works with multiple bots at once.

ViliamVadocz committed:
Add files via upload

ViliamVadocz committed:
Add files via upload

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@dataclass
+class KickoffType:
```
I'd rename it to "SpawnLocation"

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,86 @@
+@dataclass
+class KickoffExercise(TrainingExercise):
+    grader: Grader = field(default_factory=QuickResetKickoffGrader)
+    kickoffs: Union[tuple, list] = ()
```
I'd write it as `spawns: List[SpawnLocations] = field(default_factory=list)` since I don't see the value of tuples here.
note: `from typing import List`

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,86 @@
+        ex.match_config.player_configs = [
+            # Replace with path to your bot.
+            PlayerConfig.bot_config(Path(__file__).absolute(
+            ).parent.parent / 'Calculator.cfg', Team.BLUE)
```
Change it to SimpleBot like the other common_exercises

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,86 @@
+            # Replace with path to your bot.
+            PlayerConfig.bot_config(Path(__file__).absolute(
+            ).parent.parent / 'Calculator.cfg', Team.BLUE)
+        ] * len(ex.kickoffs)
```
This is creating shallow copies of one PlayerConfig. I'd suggest replacing it with `for _ in kickoffs]`

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,86 @@
+    rot: Rotator
+
+
+# Default kickoffs.
```
Maybe put these into a class or enum-class such that users of this module only need to import `Spawns` and can go `Spawns.BACK_LEFT`

DomNomNom commented on rlbottraining/common_graders/kickoff_grader.py:
```diff
@@ -0,0 +1,48 @@
+        if self.initial_seconds_elapsed is None:
+            self.initial_seconds_elapsed = game_time
+
+        ball_out_of_centre = -100 > ball.x or ball.x > 100 or -100 > ball.y or ball.y > 100
```
Maybe make this distance configurable and `sqrt(ball.x**2 + ball.y**2) > self.min_ball_displacement`. Also consider whether `ball.z` is relevant. (should we wait if the ball goes almost straight up?)

ViliamVadocz replied:
I changed it to 3d distance from start position which is simply `sqrt(ball.x**2 + ball.y**2 + (ball.z-93)**2)

DomNomNom commented on rlbottraining/common_graders/kickoff_grader.py:
```diff
@@ -0,0 +1,48 @@
+            self.initial_seconds_elapsed = game_time
+
+        ball_out_of_centre = -100 > ball.x or ball.x > 100 or -100 > ball.y or ball.y > 100
+        reset_time_reached = game_time >= self.initial_seconds_elapsed + self.reset_time
```
`min_exercise_duration` might be a better description instead of `reset_time`.
Also, is this even necessary?

ViliamVadocz replied:
Since you can't manually reset boost pads, you have to wait before a reset to make sure they're back so future kickoffs aren't off.

ViliamVadocz replied:
I found out I could lower the time. I didn't notice that before because the pads get collected when the bot gets teleported. It's not something I can really fix on my side.

ViliamVadocz committed:
Small changes based on DomNomNom's suggestions

ViliamVadocz committed:
Added match comms to let bots know it's kickoff

ViliamVadocz committed:
Added functionality for spawns on orange team.

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,122 @@
+    ]
+
+    for ex in exercises:
+        # The length of players in the match_config needs to match the number or spawns.
```
```suggestion
        # The number of players in the match_config needs to match the number of spawns.
```

DomNomNom commented on rlbottraining/common_exercises/kickoff_exercise.py:
```diff
@@ -0,0 +1,122 @@
+        assert num_players == len(self.spawns), 'Number of players does not match the number of spawns.'
+
+        car_states = {}
+        for index in range(num_players):
```
The constraint about red/orange matching the match config isn't well enforced currently.
```suggestion
        for i, player in enumerate(self.match_config.player_configs):
            assert (player.team == BLUE) == (i < len(self.blue_spawns)), "blue/orange players in matchconfig doesn't match the blue/orange spawns. expected {} blue player_configs, followed by {} orange ones.".format(len(self.blue_spawns), len(self.orange_spawns))
```

DomNomNom reviewed (changes requested):
Looks good overall.
Could you please add a test case in here to allow us to check whether it works in the future? https://github.com/RLBot/RLBotTraining/blob/master/tests/test_common_exercises.py

You can run tests using this script: https://github.com/RLBot/RLBotTraining/blob/master/test_all.bat

ViliamVadocz committed:
Enforce constraint of match config matching given spawns

Co-Authored-By: Dominik Schmid <dominikschmid93@gmail.com>

ViliamVadocz committed:
Fixed typos

Co-Authored-By: Dominik Schmid <dominikschmid93@gmail.com>

ViliamVadocz committed:
Changed string formatting to f-string

ViliamVadocz committed:
Added kickoff to test

ViliamVadocz committed:
Fixed incorrect import

DomNomNom committed:
Added a fast test for all common exercises.

DomNomNom committed:
Fix use-only-once constraint

DomNomNom reviewed (changes requested):
The test is currently failing.
To run only your test, run `python tests\test_common_exercises.py -k test_kickoff`

I've added a new test which should make it faster to iterate on trivial errors. see `test_quick.bat`
Also fixed a test which would've blocked `test_all.bat` from succeeding.

ViliamVadocz committed:
Fixed small issues that caused the test to fail

ViliamVadocz commented:
Some problems which came up:

- Importing didn't work since it's not part of the module yet. Don't know if there's something I can do.
- `Team.BLUE` did not equal `0` for some odd reason. I simply replaced it with a `0` and it worked.
- `index` was undefined. Renamed `i` to `index`.
- One too many `PassDueToBallMoved` in the test. Removed one.

DomNomNom reviewed (approved):
Sweet, looks good now. 

Thank you for your contribution!
  :Dom
